contracts/mesh-provider/src/ibc.rs

```diff
@@ -213,33 +223,110 @@ pub fn ack_list_validators(
+    let cfg = CONFIG.load(deps.storage)?;
+    let channel_id = CHANNEL.load(deps.storage)?;
+    if cfg.list_validators_retries_remaining <= 0 {
+        return Err(ContractError::NoResponse("list_validators".to_string(), channel_id))
+    }
+    let mut cfg_after = cfg.clone();
+    cfg_after.list_validators_retries_remaining -= 1;
+    CONFIG.save(deps.storage, &cfg_after)?;
+
+    // retry
+    let packet = ProviderMsg::ListValidators {};
+    let msg = IbcMsg::SendPacket {
+        channel_id: channel_id,
+        data: to_binary(&packet)?,
+        timeout: build_timeout(&env),
+    };
+    Ok(IbcBasicResponse::new().add_message(msg))
 }
 
 pub fn fail_stake(
-    _deps: DepsMut,
+    deps: DepsMut,
     // _staker is the staker Addr, not used by consumer
-    _staker: String,
-    _validator: String,
-    _amount: Uint128,
+    staker: String,
+    validator: String,
+    amount: Uint128,
 ) -> Result<IbcBasicResponse, ContractError> {
-    // TODO: release the bonded stake, adjust numer
-    unimplemented!();
+    let staker = deps.api.addr_validate(&staker)?;
+
+    if amount.is_zero() {
+        return Err(ContractError::ZeroAmount);
+    }
+
+    let mut val = VALIDATORS
+        .may_load(deps.storage, &validator)?
+        .ok_or_else(|| ContractError::UnknownValidator(validator.clone()))?;
+    let mut stake = STAKED
+        .may_load(deps.storage, (&staker, &validator))?
+        .unwrap_or_default();
+    stake.unstake_validator(&mut val, amount)?;
+    STAKED.save(deps.storage, (&staker, &validator), &stake)?;
+    VALIDATORS.save(deps.storage, &validator, &val)?;
+    Ok(IbcBasicResponse::new())
+}
+
+pub fn ack_unstake(
+    deps: DepsMut,
+    env: Env,
+    staker: String,
+    _validator: String,
+    amount: Uint128
+) -> Result<IbcBasicResponse , ContractError> {
+    let staker = deps.api.addr_validate(&staker)?;
+
+    if amount.is_zero() {
+        return Err(ContractError::ZeroAmount);
+    }
+
+    // create a future claim on number of shares (so we can adjust for later slashing)
+    let cfg = CONFIG.load(deps.storage)?;
+    let ready = env.block.time.plus_seconds(cfg.unbonding_period);
+    CLAIMS.create_claim(
+        deps.storage,
+        &staker,
+        amount,
+        Expiration::AtTime(ready),
+    )?;
+
+    // TODO: How should the user be notified of the failure?
+    Ok(IbcBasicResponse::new())
 }
 
 pub fn fail_unstake(
-    _deps: DepsMut,
+    deps: DepsMut,
     // _staker is the staker Addr, not used by consumer
-    _staker: String,
-    _validator: String,
-    _amount: Uint128,
+    staker: String,
+    validator: String,
+    amount: Uint128,
 ) -> Result<IbcBasicResponse, ContractError> {
-    // TODO: unrelease the bonded stake, remove claim
-    // Maybe we only make Claim on ack?
-    unimplemented!();
+    let staker = deps.api.addr_validate(&staker)?;
+
+    if amount.is_zero() {
+        return Err(ContractError::ZeroAmount);
+    }
+
+    // revert state changes associated with stake
```

Comment:
Wait. You no longer set the unstake in the ExecuteMsg, but only in `ack_unstake`. Then this failure wouldn't need to hold anything.

This comment you can translate to the "stake" changes. For unstake, as I mentioned above, we just do it in the ExecuteMsg handler, ack is a no-op, and fail emits an event (maybe retrying list ListValidators)

Reply:
I actually think handling unexpected failures (that should only occur on error in the consumer side, not due to possible bad arguments from the provider). I mean, retry was my first idea, but does it make sense?

Should we flag the channel as "messed up" and freeze stuff til some gov account does a reset.

This is totally out of scope of this PR and I am happy with the retry in fail_list_validators, but let's not add more retry now and think a bit larger on how to handle such situations.

Reply:
OK